Security2

VasylAzarov opened the pull request:
Implemented scripts for bd:
creating roles , user_roles tables;
inserting books, users, roles, user roles;
assign roles to users without it |
add and set spring security | add user roles | add checking role to endpoints

VasylAzarov committed:
synchronize with Security branch

VasylAzarov committed:
Implemented scripts for bd:
creating roles , user_roles tables;
inserting books, users, roles, user roles;
assign roles to users without it |
add and set spring security | add user roles | add checking role to endpoints

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/config/SecurityConfig.java:
```diff
@@ -0,0 +1,54 @@
+@EnableMethodSecurity
+@Configuration
+public class SecurityConfig {
+
+    private final CustomUserDetailsService userDetailsService;
```
```suggestion
    private final UserDetailsService userDetailsService;
```

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/config/SecurityConfig.java:
```diff
@@ -0,0 +1,54 @@
+                .authorizeHttpRequests(
+                        auth -> auth
+                                .requestMatchers(
+                                        "/auth/**",
+                                        "/error",
+                                        "/swagger-ui/**",
+                                        "/v3/api-docs/**",
+                                        "/swagger-resources/**",
+                                        "/configuration/**",
+                                        "/webjars/**")
```
```suggestion
                                        "/auth/**",
                                        "/error",
                                        "/swagger-ui/**",
                                        "/v3/api-docs/**")
```

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/controller/BookController.java:
```diff
@@ -29,45 +30,50 @@
         return bookService.findAll(pageable);
     }
 
+    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
```
in your implementation, the user with the role ADMIN has a USER role by default so you can validate only the USER role here

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/model/User.java:
```diff
@@ -0,0 +1,87 @@
+
+    private String shippingAddress;
+
+    @ManyToMany(fetch = FetchType.EAGER)
```
Don’t  use EAGER fetch type

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/model/User.java:
```diff
@@ -0,0 +1,87 @@
+@Setter
+@Entity
+@Table(name = "users")
+public class User implements UserDetails {
```
make user entity soft-deleted.

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/model/User.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
```suggestion
        return !isDeleted;
```

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/service/impl/BookServiceImpl.java:
```diff
@@ -36,6 +42,11 @@ public BookDto findById(long id) {
         return bookMapper.toDto(book);
     }
 
+    @Override
+    public boolean existByIsbn(String isbn) {
+        return bookRepository.existsByIsbn(isbn);
+    }
+
```
do you really need this method? you can call the repository method directly

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/service/impl/RoleServiceImpl.java:
```diff
@@ -0,0 +1,27 @@
+                .orElseThrow(() -> new EntityNotFoundException("Role not found: " + roleName));
+    }
+}
+
```
```suggestion
```

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,34 @@
+                    + requestDto.getEmail()
+                    + "] already exist");
+        }
+        User user = userMapper.toModel(requestDto);
```
Let’s  add the default USER role

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/sucurity/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,23 @@
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new EntityNotFoundException("Cant find user by email"));
+    }
+
```
```suggestion
```

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/sucurity/CustomUserDetailsService.java:
```diff
@@ -0,0 +1,23 @@
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new EntityNotFoundException("Cant find user by email"));
```
add email parameter

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java:
```diff
@@ -0,0 +1,40 @@
+            String password = (String) passwordField.get(value);
+            String confirmPassword = (String) confirmPasswordField.get(value);
+
+            if (Objects.equals(password, confirmPassword)) {
+                return true;
+            }
```
```suggestion
```

liliia-ponomarenko commented on src/main/java/dev/vasyl/proj/validation/PasswordMatchValidator.java:
```diff
@@ -0,0 +1,40 @@
+            if (Objects.equals(password, confirmPassword)) {
+                return true;
+            }
+            return password.equals(confirmPassword);
```
```suggestion
            return Objects.equals(password, confirmPassword);
```

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/04-create-user_roles-table.yaml:
```diff
@@ -0,0 +1,31 @@
+      runOnChange: true
+      changes:
+        - createTable:
+            tableName: user_roles
```
```suggestion
            tableName: users_roles
```

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/05-insert-books-if-not-exist.yaml:
```diff
@@ -0,0 +1,76 @@
+      id: insert-books-if-not-exist
+      author: vasyl
+      runOnChange: true
+      preConditions:
+        - onFail: MARK_RAN
+          sqlCheck:
+            sql: "SELECT COUNT(*) FROM books"
+            expectedResult: 0
```
do you really need this check? Remove or refactor it;)

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/07-insert-users-if-not-exist.yaml:
```diff
@@ -0,0 +1,67 @@
+      preConditions:
+        - onFail: MARK_RAN
+          sqlCheck:
+            sql: "SELECT COUNT(*) FROM users"
```
the same. select by email or remove this condition

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/08-insert-admin-user-if-none-exist.yaml:
```diff
@@ -0,0 +1,45 @@
+        - sqlCheck:
+            expectedResult: 0
+            sql: |
+              SELECT COUNT(*)
+              FROM user_roles ur
+              JOIN roles r ON ur.role_id = r.id
+              WHERE r.name = 'ADMIN'
```
the same issue

liliia-ponomarenko commented on src/main/resources/db/changelog/changes/09-assign-user-role-to-users-without-roles.yaml:
```diff
@@ -0,0 +1,20 @@
+databaseChangeLog:
+  - changeSet:
+      id: assign-user-role-to-users-without-roles
+      author: vasyl
+      runOnChange: true
+      preConditions:
+        - onFail: MARK_RAN
+      changes:
+        - sql:
+            sql: |
+              INSERT INTO user_roles (user_id, role_id)
+              SELECT u.id, r.id
+              FROM users u
+              CROSS JOIN roles r
+              WHERE r.name = 'USER'
+              AND NOT EXISTS (
+                SELECT 1
+                FROM user_roles ur
+                WHERE ur.user_id = u.id
+              );
```
Let’s remove it.

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s  improve your solution!

VasylAzarov committed:
refactored yaml scripts |
refactored role check in controller end security |
add soft delete for User entity |
add default User role for new Users
| replace fetch type EAGER with EntityGraph in User entity

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/model/Role.java:
```diff
@@ -0,0 +1,26 @@
+@Table(name = "roles")
+@Getter
+@Setter
+public class Role {
```
```suggestion
public class Role implements GrantedAuthority {
```

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/model/User.java:
```diff
@@ -0,0 +1,93 @@
+@Setter
+@Entity
+@SQLDelete(sql = "UPDATE users SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
check if `@Where` is deprecated or not

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/model/User.java:
```diff
@@ -0,0 +1,93 @@
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id")
+    )
+    private Set<Role> roles;
```
```suggestion
    private Set<Role> roles = new HashSet<>();
```

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/model/User.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Override
+    public Collection<? extends GrantedAuthority> getAuthorities() {
+        return List.of();
```
```suggestion
        return roles;
```

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/repository/UserRepository.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Be consistent, if you use `@Repository` use it everywhere or remove

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/service/RoleService.java:
```diff
@@ -0,0 +1,10 @@
+import dev.vasyl.proj.model.Role;
+import dev.vasyl.proj.model.RoleName;
+
+public interface RoleService {
```
remove this service

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/service/impl/RoleServiceImpl.java:
```diff
@@ -0,0 +1,26 @@
+
+@Service
+@RequiredArgsConstructor
+public class RoleServiceImpl implements RoleService {
```
remove

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/service/impl/UserServiceImpl.java:
```diff
@@ -0,0 +1,43 @@
+        Role role = roleRepository.findByName(RoleName.USER).orElseThrow(
+                () -> new EntityNotFoundException("Error when set user role"));
+        user.setRoles(Set.of(role));
+        user = userRepository.save(user);
+        return userMapper.toUserResponse(user);
```
```suggestion
        userRepository.save(user);
        return userMapper.toUserResponse(user);
```

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java:
```diff
@@ -0,0 +1,41 @@
+
+@Service
+@RequiredArgsConstructor
+public class UserDetailsService
```
```suggestion
public class CustomUserDetailsService implements UserDetailsService {
```

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java:
```diff
@@ -0,0 +1,41 @@
+package dev.vasyl.proj.sucurity;
```
```suggestion
package dev.vasyl.proj.security;
```

Elena-Bruyako commented on src/main/java/dev/vasyl/proj/sucurity/UserDetailsService.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .map(user -> {
+                    List<GrantedAuthority> authorities = user.getRoles().stream()
+                            .map(role -> new SimpleGrantedAuthority("ROLE_" + role.getName()))
+                            .collect(Collectors.toList());
+                    if (authorities.contains(new SimpleGrantedAuthority("ROLE_ADMIN"))) {
+                        authorities.addAll(List.of(
+                                new SimpleGrantedAuthority("ROLE_USER")
+                        ));
+                    }
+                    return new org.springframework.security.core.userdetails.User(
+                            user.getEmail(),
+                            user.getPassword(),
+                            authorities
+                    );
+                })
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Cant find user by email: [" + email + "]"));
```
```suggestion
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new EntityNotFoundException(.......
```

Elena-Bruyako commented on src/main/resources/db/changelog/changes/05-insert-books.yaml:
```diff
@@ -0,0 +1,96 @@
+databaseChangeLog:
+  - changeSet:
+      id: insert-books-1
```
don't need insert books to DB

Elena-Bruyako reviewed (changes requested):
Let's resolve conflicts in branch

see comments

VasylAzarov committed:
removed unnecessary RoleService, RoleServiceImpl, 05-insert-books.yaml, again renamed UserDetailsService -> CustomUserDetailsService | add  GrantedAuthority implementation for Role GrantedAuthority | and some minor fixes

VasylAzarov committed:
Merge branch 'main' into Security2

sarakhmen reviewed (approved):
lgtm!
